adr/ADR-19.md

```diff
@@ -0,0 +1,35 @@
+# JetStream Automatic Status Management
+
+|Metadata|Value|
+|--------|-----|
+|Date    |2021-09-20|
+|Author  |@scottf|
+|Status  |Partially Implemented|
+|Tags    |jetstream, client|
+
+## Context
+
+This document attempts to describe the Automatic Status Management (ASM) tasks to be performed by the client between receipt of messages and hand-off to the client user.
+
+ASM should be the default client behavior. Optionally, clients can allow users to choose to handle the status messages themself.
+
+## General
+
+Intercept all status messages, never pass to user. If there is some exceptional reason to notify or escalate some state to the user, user a language specific paradigm such as an exception.
+
+## Flow Control and Heartbeat
+
+1. Automatically respond to flow control messages as soon as it is observed in the normal processing of next message. Escalate to user when unable to respond to the flow control.
```

Comment:
Maybe I misread what you wrote here, but when receiving a Flow Control, the library needs to respond *only* after all pending messages have been processed. Otherwise, this defeats the point of FC.

Reply:
"as soon as it is observed"

Reply:
I am suggesting that it's processed as the user would normally get the message. I would argue against processing it when is received over the network. If the user isn't processing messages, it makes sense that they are not ready to resume getting messages from the server.

Reply:
Back to my example. I get 350 messages, then I get the FC, but the server has already sent me another 300 or so messages and they are buffered. I would not hold the fc until the buffer is empty.

Reply:
Well no, the server will not send you another 300 messages if it has sent an FC to which the library did not respond to. But as long as we are clear that the library should not simply reply to an FC message when it gets it, but instead when all previous messages have been delivered, then it is ok.

Reply:
I think we are saying the same thing. There is a difference between the client buffering the messages the server has sent and the user asking for it. It doesn't matter if the FC is in the buffer, it is only responded to when the user asks for the next message and it turns out to be the fc.

Reply:
Right